// sudoku-core/dist/cjs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.hint = exports.solve = exports.generate = exports.analyze = void 0;
const sudoku_1 = require("./sudoku");
const sudoku_solver_1 = require("./sudoku-solver");
function analyze(Board) {
    const { analyzeBoard } = (0, sudoku_1.createSudokuInstance)({
        initBoard: Board.slice(),
    });
    return Object.assign(Object.assign({}, analyzeBoard()), { hasUniqueSolution: (0, sudoku_solver_1.isUniqueSolution)(Board) });
}
exports.analyze = analyze;
function generate(difficulty) {
    const { getBoard } = (0, sudoku_1.createSudokuInstance)({ difficulty });
    if (!analyze(getBoard()).hasUniqueSolution) {
        return generate(difficulty);
    }
    return getBoard();
}
exports.generate = generate;
function solve(Board) {
    const solvingSteps = [];
    const { solveAll } = (0, sudoku_1.createSudokuInstance)({
        initBoard: Board.slice(),
        onUpdate: (solvingStep) => solvingSteps.push(solvingStep),
    });
    const analysis = analyze(Board);
    if (!analysis.hasSolution) {
        return { solved: false, error: "No solution for provided board!" };
    }
    const board = solveAll();
    if (!analysis.hasUniqueSolution) {
        return {
            solved: true,
            board,
            steps: solvingSteps,
            analysis,
            error: "No unique solution for provided board!",
        };
    }
    return { solved: true, board, steps: solvingSteps, analysis };
}
exports.solve = solve;
function hint(Board) {
    const solvingSteps = [];
    const { solveStep } = (0, sudoku_1.createSudokuInstance)({
        initBoard: Board.slice(),
        onUpdate: (solvingStep) => solvingSteps.push(solvingStep),
    });
    const analysis = analyze(Board);
    if (!analysis.hasSolution) {
        return { solved: false, error: "No solution for provided board!" };
    }
    const board = solveStep();
    if (!board) {
        return { solved: false, error: "No solution for provided board!" };
    }
    if (!analysis.hasUniqueSolution) {
        return {
            solved: true,
            board,
            steps: solvingSteps,
            analysis,
            error: "No unique solution for provided board!",
        };
    }
    return { solved: true, board, steps: solvingSteps, analysis };
}
exports.hint = hint;
